api/src/main/java/jakarta/mail/Message.java

```diff
@@ -705,4 +707,33 @@ protected void setExpunged(boolean expunged) {
     public boolean match(SearchTerm term) throws MessagingException {
         return term.match(this);
     }
+
+    /**
+     * Obtains the {@link StreamProvider} from the session, if exists.
+     * Otherwise it obtains it from
+     * {@link Session#getDefaultInstance(java.util.Properties, Authenticator)}.
+     *
+     * @return the StreamProvider implementation from the session.
+     * @throws MessagingException if errors.
+     *
+     * @since JavaMail 2.2
+     */
+    protected StreamProvider provider() throws MessagingException {
+        try {
+            try {
+                final Session s = this.session;
+                if (s != null) {
+                    return s.getStreamProvider();
+                } else {
+                    return Session.getDefaultInstance(System.getProperties(),
```

Comment:
We don't have to solve this point in this ticket but, you made a good point about default instance being static which means we could consider using `return StreamProvider.provider()`.  My original thinking was to use the default session to match what was done in Transport and to avoid paying the lookup over and over.

Reply:
I was re-thinking it and what makes more sense to me is:

If the user wants to use its custom StreamProvider, he has to pass the Session in the Message. If he does not pass the session is returns what the default session instance has.

Otherwise it will suffer the performance.